File: Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        if not nums: return [[]]
        res = [[]]
        tmp = []
        self.helper(nums,0,res,tmp)
        return res
    
    def helper(self,nums,start,res,tmp):
        for i in xrange(start,len(nums)):
            tmp.append(nums[i])
            res.append(list(tmp))
            self.helper(nums,i + 1, res,tmp)
            tmp.pop()
        return

File: subsets.kt
class Solution {
    fun subsets(nums: IntArray): List<List<Int>> {
        val tmp = mutableListOf<Int>()
        val result = MutableList<List<Int>>(1){listOf<Int>()}
        // result.add(listOf())
        fun dfs(index: Int) {
            for(i in index until nums.size){
                tmp.add(nums[i])
                result.add(tmp.toList())
                dfs(i+1)
                tmp.removeLast()
            }
        }
        dfs(0)
        return result
    }
}
